pkg/frontend/querymiddleware/roundtrip_test.go

```diff
@@ -794,6 +800,167 @@ func TestTripperware_RemoteRead(t *testing.T) {
+				return httptest.NewRequest("GET", cardinalityLabelValuesPathSuffix, nil)
+			},
+		},
+		"cardinality active series": {
+			makeRequest: func() *http.Request {
+				return httptest.NewRequest("GET", cardinalityActiveSeriesPathSuffix, nil)
+			},
+		},
+		"cardinality active native histograms": {
+			makeRequest: func() *http.Request {
+				return httptest.NewRequest("GET", cardinalityActiveNativeHistogramMetricsPathSuffix, nil)
+			},
+		},
+		"label names": {
+			makeRequest: func() *http.Request {
+				return httptest.NewRequest("GET", labelNamesPathSuffix, nil)
+			},
+		},
+		"remote read": {
+			makeRequest: func() *http.Request {
+				return makeTestHTTPRequestFromRemoteRead(&prompb.ReadRequest{
+					Queries: []*prompb.Query{
+						{
+							Matchers:         []*prompb.LabelMatcher{{Name: "__name__", Type: prompb.LabelMatcher_EQ, Value: "some_metric"}},
+							StartTimestampMs: time.Now().Add(-60 * 24 * time.Hour).UnixMilli(),
+							EndTimestampMs:   time.Now().UnixMilli(),
+						},
+					},
+				})
+			},
+		},
+	}
+
+	ctx := context.Background()
+	logger := log.NewNopLogger()
+
+	// Setup a fake Kafka cluster.
+	_, clusterAddr := testkafka.CreateCluster(t, numPartitions, topic)
+
+	// Write some records to different partitions.
+	expectedOffsets := produceKafkaRecords(t, clusterAddr, topic,
+		&kgo.Record{Partition: 0},
+		&kgo.Record{Partition: 0},
+		&kgo.Record{Partition: 0},
+		&kgo.Record{Partition: 1},
+		&kgo.Record{Partition: 1},
+		&kgo.Record{Partition: 2},
+	)
+
+	// Create the topic offsets reader.
+	readClient, err := ingest.NewKafkaReaderClient(createKafkaConfig(clusterAddr, topic), nil, logger)
+	require.NoError(t, err)
+	t.Cleanup(readClient.Close)
+
+	offsetsReader := ingest.NewTopicOffsetsReader(readClient, topic, 100*time.Millisecond, nil, logger)
+	require.NoError(t, services.StartAndAwaitRunning(ctx, offsetsReader))
+	t.Cleanup(func() {
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, offsetsReader))
+	})
+
+	// Create the tripperware.
+	tw, err := NewTripperware(
+		makeTestConfig(func(cfg *Config) {
+			cfg.ShardedQueries = false
+			cfg.SplitQueriesByInterval = 0
+			cfg.CacheResults = false
+		}),
+		log.NewNopLogger(),
+		mockLimits{},
+		NewPrometheusCodec(nil, 0, formatJSON),
+		nil,
+		promql.EngineOpts{
+			Logger:     log.NewNopLogger(),
+			Reg:        nil,
+			MaxSamples: 1000,
+			Timeout:    time.Minute,
+		},
+		true,
+		offsetsReader,
+		nil,
+	)
+	require.NoError(t, err)
+
+	// Test it against all routes.
+	for testName, testData := range tests {
+		testData := testData
+
+		t.Run(testName, func(t *testing.T) {
+			t.Parallel()
+
+			for _, consistencyLevel := range []string{querierapi.ReadConsistencyEventual, querierapi.ReadConsistencyStrong} {
+				consistencyLevel := consistencyLevel
+
+				t.Run(fmt.Sprintf("consistency level: %s", consistencyLevel), func(t *testing.T) {
+					// Create a roundtripper that captures the downstream HTTP request.
```

Comment:
```suggestion
					t.Parallel()
					// Create a roundtripper that captures the downstream HTTP request.
```

maybe in parallel? I see the tripperware is shared between tests, but it should be concurrency-safe, right?

Reply:
Yes, that was the plan, but I missed it.

Reply:
Done in https://github.com/grafana/mimir/pull/8808/commits/94d767c523f97d70cffbef4f5f3a3e574b6674e9